### ISSUE TITLE
[Generics] Code generation is failing when parameters are immutable generics

### ISSUE THREAD

**Issue by cschaller1** [immutables/immutables#548]
Here is an example :

```java
abstract class AbstractAClass<T>  {
	@Value.Parameter
	public abstract BClass<T> getValue(); // note BClass is also generated by Immutables
}
```
Then the generation is failing and we get stuff like this : 

```java
public final class AClass<T> extends AbstractAClass<T> {
  // compilation error with <any>
  private final <any> value;
 }
```

**Comment by elucash** [comment]
This is a known limitation. Compilers are helpless here, both Javac and ECJ have no clue how to treat yet-to-be-generated parametrized types. Last time I've looked into it I was trying to parse the type from the source code, but for some reason (I believe there were quite important reasons) I gave up on this. I'll circle back on this again, but no promises can be made about when. In a mean time what I did and suggest other people to avoid using inverse generation style (`AbstractFoo -> Foo`), but instead use sandwich pattern illustrated in the example at the top of the readme file: https://github.com/immutables/immutables/blob/master/README.md

**Comment by io7m** [comment]
I've just run into this one too. I was going to report it but discovered this existing bug.

Here's a repro case, for future reference:

https://github.com/io7m/immutables-bug-20170227

**Comment by TimW2014** [comment]
Interestingly, under Eclipse, annotation processing works correctly with generics: 

@Nullable ArtifactSummary<PresalesSurveyDto> getArtifact(); // becomes...
 private final @Nullable ArtifactSummary<PresalesSurveyDto> artifact;

Yet when run via Maven, we're seeing (as noted above):

private final @Nullable <any> artifact;

What's the difference?  Does Eclipse feed the annotation process more detailed information?

Side note: Great package!  But inability to handle any generics (in the most commonly-recommended pattern) is fairly serious! It should be noted on the introductory material or documentation, IMO, as well explaining the workaround.

**Comment by TimW2014** [comment]
Hmm... looks like the "sandwich" workaround produces the same results.

**Comment by elucash** [comment]
Unfortunately, Javac does not return any meaningful information when not-yet-generated-type contains generics. ECJ(Eclipse) compiler is better in this regard. Reading types from the source. I do feel "sandwich" will work, the trick is to not use not-yet-generated-types as attribute types. We're still investigating if we can read this info from the source files (which is not trivial on it's own), but the source information not always available, for example, during incremental compilation.

**Comment by TimW2014** [comment]
Ah; this only happens when the both classes (the one referenced by the generic parameter & its declaring class) are being generated by Immutables in the same project (hence explaining why it works for, say, Optional), and the 'sandwich' has to be applied to the *referenced* class, not the one in which the parameter is being declared.

That's good enough for me, for now.  Thank you very much!

**Comment by elucash** [comment]
JFYI master contains a likely fix (#725) for this issue to be released soon in v2.6
